avogadro/calc/chargemanager.cpp

```diff
@@ -0,0 +1,185 @@
+/******************************************************************************
+  This source file is part of the Avogadro project.
+  This source code is released under the 3-Clause BSD License, (see "LICENSE").
+******************************************************************************/
+
+#include "chargemanager.h"
+#include "chargemodel.h"
+#include "defaultmodel.h"
+
+#include <algorithm>
+#include <memory>
+
+using std::unique_ptr;
+
+namespace Avogadro {
+namespace Calc {
+
+ChargeManager& ChargeManager::instance()
+{
+  static ChargeManager instance;
+  return instance;
+}
+
+void ChargeManager::appendError(const std::string& errorMessage)
+{
+  m_error += errorMessage + "\n";
+}
+
+bool ChargeManager::registerModel(ChargeModel* model)
+{
+  return instance().addModel(model);
+}
+
+bool ChargeManager::unregisterModel(const std::string& identifier)
+{
+  return instance().removeModel(identifier);
+}
+
+bool ChargeManager::addModel(ChargeModel* model)
+{
+  if (model == nullptr) {
+    appendError("Supplied model was null.");
+    return false;
+  }
+
+  if (m_identifiers.find(model->identifier()) != m_identifiers.end()) {
+    appendError("Model " + model->identifier() + " already loaded.");
+    return false;
+  }
+  for (std::vector<ChargeModel*>::const_iterator it = m_models.begin();
+       it != m_models.end(); ++it) {
+    if (*it == model) {
```

Comment:
This was copy / paste from the same code in file formats. But it's possible to have different ways to load a PDB or XYZ file (e.g., through our code or Open Babel).

I'm not going to let there be multiple charge models with the same identifier -- it would be too confusing.